src/MessagePack/BufferWriter.cs

```diff
@@ -0,0 +1,192 @@
+
+using System;
+using System.Buffers;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace MessagePack
+{
+    /// <summary>
+    /// A fast access struct that wraps <see cref="IBufferWriter{T}"/>.
+    /// </summary>
+    internal ref struct BufferWriter
+    {
+        /// <summary>
+        /// The underlying <see cref="IBufferWriter{T}"/>.
+        /// </summary>
+        private IBufferWriter<byte> _output;
+
+        /// <summary>
+        /// The result of the last call to <see cref="IBufferWriter{T}.GetSpan(int)"/>, less any bytes already "consumed" with <see cref="Advance(int)"/>.
+        /// Backing field for the <see cref="Span"/> property.
+        /// </summary>
+        private Span<byte> _span;
+
+        /// <summary>
+        /// The result of the last call to <see cref="IBufferWriter{T}.GetMemory(int)"/>, less any bytes already "consumed" with <see cref="Advance(int)"/>.
+        /// </summary>
+        private ArraySegment<byte> _segment;
+
+        /// <summary>
+        /// The number of uncommitted bytes (all the calls to <see cref="Advance(int)"/> since the last call to <see cref="Commit"/>).
+        /// </summary>
+        private int _buffered;
+
+        /// <summary>
+        /// The total number of bytes written with this writer.
+        /// Backing field for the <see cref="BytesCommitted"/> property.
+        /// </summary>
+        private long _bytesCommitted;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BufferWriter"/> struct.
+        /// </summary>
+        /// <param name="output">The <see cref="IBufferWriter{T}"/> to be wrapped.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public BufferWriter(IBufferWriter<byte> output)
+        {
+            _buffered = 0;
+            _bytesCommitted = 0;
+            _output = output;
+
+            var memory = _output.GetMemory();
+            MemoryMarshal.TryGetArray(memory, out _segment);
+            _span = memory.Span;
+        }
+
+        /// <summary>
+        /// Gets the result of the last call to <see cref="IBufferWriter{T}.GetSpan(int)"/>.
+        /// </summary>
+        public Span<byte> Span => _span;
+
+        /// <summary>
+        /// Gets the total number of bytes written with this writer.
+        /// </summary>
+        public long BytesCommitted => _bytesCommitted;
+
+        /// <summary>
+        /// Gets the <see cref="IBufferWriter{T}"/> underlying this instance.
+        /// </summary>
+        internal IBufferWriter<byte> UnderlyingWriter => _output;
+
+        public Span<byte> GetSpan(int sizeHint)
+        {
+            Ensure(sizeHint);
+            return this.Span;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ref byte GetPointer(int sizeHint)
+        {
+            Ensure(sizeHint);
+
+            if (_segment.Array != null)
```

Comment:
I fear this runtime check will miss out on future .NET Core optimizations.

Reply:
Do you have any suggestions? Regardless of future optimizations, this change was instrumental in dramatically improving performance in  net47, and had no negative netcore impact.

Reply:
It’s always using the array. I would ifdef for NETCore and skip the GetMemory calls altogether

Reply:
I just tried that. .NET Core gets 24% _slower_ when I remove this ArraySegment optimization. So I'm keeping it for now.